### PR TITLE
304906 improve pre tooltips 2

### PR DESCRIPTION
**Pull request by blackears**
Resolves: *(direct link to the issue)*

*(short description of the changes and the motivation to make the changes)*

- [ ] I signed [CLA](https://musescore.org/en/cla)
- [ ] I made sure the code in the PR follows [the coding rules](https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules)
- [ ] I made sure the code compiles on my machine
- [ ] I made sure there are no unnecessary changes in the code
- [ ] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [ ] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [ ] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

**Comment by blackears** [comment]
I do not know why this is touching so many files.  I've made these changes on top of a fresh branch of 3.x on my local machine.

**Comment by Marr11317** [comment]
because you should make the pr target the 3.x branch, not the master branch

**Comment by Jojo-Schmitz** [comment]
Try `git pull --rebase musescore 3.x`, if musescore is the name of your upstream repo

File: mscore/pianoroll/pianolevelsfilter.cpp [hunk 1]
```diff
@@ -11,19 +11,19 @@ namespace Ms {
 
 
 static const char* STRN_NOTE_ON_NAME = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Note start time");
-static const char* STRN_NOTE_ON_TT = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Add (or subtract) a bit to the start time of a note");
+static const char* STRN_NOTE_ON_TT = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Add (or subtract) 1/1000 of note's length to the start time of a note");
 
 static const char* STRN_LEN_MUL_NAME = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Length (multiplier)");
-static const char* STRN_LEN_MUL_TT = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Multiply the note length by a bit");
+static const char* STRN_LEN_MUL_TT = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Remove 1/1000 of note's length to the end of a note");
```
Comment (Jojo-Schmitz) [inline mscore/pianoroll/pianolevelsfilter.cpp:17]: not `at the end of a note`?

File: mscore/pianoroll/pianolevelsfilter.cpp [hunk 2]
```diff
@@ -11,19 +11,19 @@ namespace Ms {
 
 
 static const char* STRN_NOTE_ON_NAME = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Note start time");
-static const char* STRN_NOTE_ON_TT = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Add (or subtract) a bit to the start time of a note");
+static const char* STRN_NOTE_ON_TT = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Add (or subtract) 1/1000 of note's length to the start time of a note");
```
Comment (Jojo-Schmitz) [inline mscore/pianoroll/pianolevelsfilter.cpp:14]: `to (or from)`?

File: mscore/pianoroll/pianolevelsfilter.cpp [hunk 3]
```diff
@@ -11,19 +11,19 @@ namespace Ms {
...
 
 static const char* STRN_LEN_MUL_NAME = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Length (multiplier)");
-static const char* STRN_LEN_MUL_TT = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Multiply the note length by a bit");
+static const char* STRN_LEN_MUL_TT = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Remove 1/1000 of note's length to the end of a note");
 
 static const char* STRN_LEN_OFF_NAME = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Length (offset)");
-static const char* STRN_LEN_OFF_TT = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Add (or subtract) a bit from the end of the note");
+static const char* STRN_LEN_OFF_TT = QT_TRANSLATE_NOOP("PianoLevelsFilter", "Remove 1/1000 of whole note from the end of the note");
```
Comment (Jojo-Schmitz) [inline mscore/pianoroll/pianolevelsfilter.cpp:20]: no substract anymore?

Comment (blackears) [reply mscore/pianoroll/pianolevelsfilter.cpp:20]: No.  Just subtraction.

File: mscore/pianoroll/pianolevelsfilter.cpp [hunk 4]
```diff
@@ -140,37 +140,19 @@ QString PianoLevelFilterLenOfftime::tooltip()
...
-      int ticks = chord->ticks().ticks();
-      int gate = evt->len();
-      int offTicks = ticks - (ticks * gate / 1000);
+            Chord* chord = note->chord();
+            Fraction noteLen = chord->ticks();
+            int evtLen = evt->len();
+            Fraction offsetLen = noteLen - (noteLen * evtLen / 1000);
 
-      return offTicks;
+            return offsetLen.numerator() * 1000 / offsetLen.denominator();
```
Comment (Jojo-Schmitz) [inline mscore/pianoroll/pianolevelsfilter.cpp:155]: Indent is wrong

File: mscore/pianoroll/pianolevelsfilter.cpp [hunk 5]
```diff
@@ -179,19 +161,21 @@ int PianoLevelFilterLenOfftime::value(Staff* /*staff*/, Note* note, NoteEvent* e
...
-      ne.setLen(gate);
+            NoteEvent ne = *evt;
+            ne.setLen(evtLen);
 
-      score->startCmd();
-      score->undo(new ChangeNoteEvent(note, evt, ne));
-      score->endCmd();
+            score->startCmd();
+            score->undo(new ChangeNoteEvent(note, evt, ne));
+            score->endCmd();
```
Comment (Jojo-Schmitz) [inline mscore/pianoroll/pianolevelsfilter.cpp:178]: here too

File: libmscore/fraction.h [hunk 6]
```diff
@@ -188,20 +188,26 @@ class Fraction {
...
 
-      #if 0
+
       Fraction& operator/=(int val)
             {
             _denominator *= val;
+            if (_denominator < 0) {
+                  _denominator = -_denominator;
+                  _numerator = -_numerator;
+            }
```
Comment (Jojo-Schmitz) [inline libmscore/fraction.h:198]: Indent, that one brace ;-)

**Comment by Jojo-Schmitz** [comment]
Don't quite understand where all these other commits stem from, probably because you created this PR against master rather than 3.x?

**Comment by blackears** [comment]
I created #6040 against master, but had similar problems so I created this one off of a fresh copy of 3.x.  I only copied across the piano* files from #6040.  There shouldn't be anything related to master here.

**Comment by Jojo-Schmitz** [comment]
Not sure it is possible to correct this, so you may need to create a new branch off of 3.x and `git cherry-pick 12269f6` this commit into that, close this PR and send a new one?

**Comment by blackears** [comment]
```git diff 3.x``` shows me only the files' I've modified.

**Comment by blackears** [comment]
So do this?:
```git checkout 3.x
git checkout -b 304906-improve-pre-tooltips-3
git cherry-pick 0ba1952```

**Comment by Jojo-Schmitz** [comment]
> `git diff 3.x` shows me only the files' I've modified.

Still your PR is against master, not against 3.x

**Comment by blackears** [comment]
How do I switch to using 3.x?

**Comment by Jojo-Schmitz** [comment]
I don't think you can on an existing PR.

**Comment by Jojo-Schmitz** [comment]
Continued in #6043

**Pull request closed without merging**